Client/mods/deathmatch/logic/lua/CLuaArguments.cpp

```diff
@@ -535,46 +535,51 @@ json_object* CLuaArguments::WriteTableToJSONObject(bool bSerialize, CFastHashMap
 
     pKnownTables->insert(std::make_pair(this, pKnownTables->size()));
 
-    bool                                  bIsArray = true;
-    unsigned int                          iArrayPos = 1;            // lua arrays are 1 based
-    vector<CLuaArgument*>::const_iterator iter = m_Arguments.begin();
+    bool                                      bIsArray = true;
+    vector<pair<unsigned int, CLuaArgument*>> vecSortedArguments;            // lua arrays are not necessarily sorted
+    vector<CLuaArgument*>::const_iterator     iter = m_Arguments.begin();
     for (; iter != m_Arguments.end(); iter += 2)
     {
         CLuaArgument* pArgument = *iter;
         if (pArgument->GetType() == LUA_TNUMBER)
         {
             double       num = pArgument->GetNumber();
             unsigned int iNum = static_cast<unsigned int>(num);
-            if (num == iNum)
-            {
-                if (iArrayPos != iNum)            // check if the value matches its index in the table
-                {
-                    bIsArray = false;
-                    break;
-                }
-            }
-            else
-            {
-                bIsArray = false;
-                break;
-            }
+
+            vecSortedArguments.push_back({iNum, *(iter + 1)});
         }
         else
         {
             bIsArray = false;
             break;
         }
-        iArrayPos++;
     }
 
-    if (bIsArray)
+    if (bIsArray)            // the table could possibly be an array
     {
-        json_object*                          my_array = json_object_new_array();
-        vector<CLuaArgument*>::const_iterator iter = m_Arguments.begin();
-        for (; iter != m_Arguments.end(); iter++)
+        // sort the table based on the keys (already handled correctly by std::pair)
+        std::sort(vecSortedArguments.begin(), vecSortedArguments.end());
+
+        // only the first and last element are checked, everything else is correct by default because the vector was sorted
+        unsigned int const iFirstKey = vecSortedArguments.front().first;
```

Comment:
What if `vecSortedArguments` is empty?

Reply:
That's an oversight on my part, I used to have everything inline in the if-statement which worked because of lazy evaluation. Thanks for the catch!